design simplicity based on part classes

agrrr3 opened the pull request:
Works, and almost ready

~~this should work but it does not.~~

~~If I build my robotics, the sitrep reports the values just fine e.g. for a robotic with mass driver, flak and armor: 2 part type classes, 0.8 factor, and current policy factor 0.8 or 1.0 depending if you applied the policy or not . If you go MD, armor, detection, a fuel and a troop part, the factor with policy applied goes up to 1.0~~

~~I also see difference from applying the policy in the ship designer and the production queue. There the five part class design costs with the simplicity policy adopted only 62.2 PP  instead of 77.7PP~~

~~Buuuuut, there is a difference between what the ships do sitrep and what the actual cost in design/production is - in design/production the 0.8 factor always seems to get triggered even if there are many part class in the ship design.~~


[forum discussion](https://freeorion.org/forum/viewtopic.php?p=122421#p122421)

done:

- add pedia descriptions
- check AI for simplicity usage etc
- cleanup code and history
- make linting happy
- production cost in design and actual production

o01eg commented:
I wonder if any new things in Python FOCS should be introduced in modules.

o01eg commented:
> Not sure what the structure should be/how to cut.

@agrrr3  I think to put `DesignHasPartClass` condition into new `_conditions` module and `PartOfClassInShipDesign` into `_value_refs`.

And any new `py::class_` definitions should be in their own module too as well as enum values too.

agrrr3 commented:
> > focs currently lacks (e.g manipulating/querying list properties).
> 
> Could you explain what that mwans, perhaps with a small example of use?

Well, we have Statistics to handle e.g. counting (measuring a an arbitrarily filtered set of universe objects into a vref). But for example in this PR i have to check for each of the part classes in an explicit statement and add that up instead of counting the unique part classes. If we add another part class we have to change this code, because we need to add a new statement with the new part class name.

Had similar issues with tags, specials...

agrrr3 commented:
I did a prototypical implementation in 6fbfb30d1718016caa9350c54b3b437ec8f11f5c
to see what is necessary to make this work.

I got a simple pedia view working (that is single-click a ship design in ship design UI works; double-click, which opens the layout does show the wrong values/the values based on empire source). It made more cases necessary where I use the temp ship. Probably at least two places are needed where the temp ship gets tweaked. Also showing monsters in the single-click view is broken, as i set the owner to the empire and the empire policies get applied.

In this count-the-distinct-part-classes case (and i guess in most cases) no ship is really necessary, as the info is not dependent on ship state but only on the design. So maybe adding a lot of boilerplate valuerefs for querying ship designs is better.

geoffthemedio commented:
> if i could access the ship part classes as vrefs; (probably map/reduce style), i probably would get around using the scoping issue which currently happens with conditions.

It should be possible to add a way to get either a list of strings or some information about the list of strings to suit what you want.

A question is perhaps how flexible the implementation should be at this point...

Something that does specifically what you need would be easier to implement, but wouldn't be composable / resuable for other scripting goals. For example, a new ship property could just directly give how many unique (non-empty) part classes that ship's design uses. That would be relatively easy to set up. Or a `ComplexVariable` that takes a ship design ID and returns the number of unique part classes / types it has.

Or, there is [internal engine support](https://github.com/freeorion/freeorion/blob/master/universe/ValueRefs.cpp#L1296) for `ValueRef<vector<string>>`. This can already get a list of names of parts in a ship's design, though there's no currently parsing support and no way to do anything with that list for game-mechanics purposes (the feature is used in the UI, though, in particular object list columns for multi-valued properties of objects). But there could additional be ways to manipulate a `ValueRef<vector<string>>` result in FOCS expressions. One could potentially get a list of part names for an object or for a ship design id, and calculate the number of unique values in that list, or other similar results / statistics / operations. That would probably need some new type of FOCS expression node that takes a `ValueRef<vector<string>>` and evaulates it and returns information from or about the result, similar to what `Statistic` does for a `ValueRef<string>` evaluated on multiple objects. Doing this would be substantially more effort.

geoffthemedio commented on universe/ValueRefs.cpp:
```diff
@@ -2806,6 +2806,36 @@ std::vector<std::string> ComplexVariable<std::vector<std::string>>::Eval(
+
+            const ShipDesign* design = context.ContextUniverse().GetShipDesign(design_id);
+            if (!design)
+	      return {};
```
tabs to spaces

geoffthemedio commented on universe/ValueRefs.cpp:
```diff
@@ -2806,6 +2806,36 @@ std::vector<std::string> ComplexVariable<std::vector<std::string>>::Eval(
         return std::vector<std::string>{pols.begin(), pols.end()};
     }
+  //    IF_CURRENT_VALUE(std::vector<std::string>) //XXX
+       else if (m_property_name == "PartClassesInShipDesign") {
```
````
 universe/ValueRefs.cpp | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/universe/ValueRefs.cpp b/universe/ValueRefs.cpp
index 4cde4e164..83570ce8c 100644
--- a/universe/ValueRefs.cpp
+++ b/universe/ValueRefs.cpp
@@ -1981,6 +1981,20 @@ int ComplexVariable<int>::Eval(const ScriptingContext& context) const
         };
         return static_cast<int>(range_count_if(design->Parts(), part_of_class));
     }
+    else if (m_property_name == "NumPartClassesInShipDesign") {
+        if (!m_int_ref1)
+            return 0;
+
+        const int design_id = m_int_ref1->Eval(context);
+        if (design_id == INVALID_DESIGN_ID)
+            return 0;
+
+        const ShipDesign* design = context.ContextUniverse().GetShipDesign(design_id);
+        if (!design)
+            return 0;
+
+        return design->PartClassCount().size();
+    }
     else if (m_property_name == "JumpsBetween") {
         int object1_id = INVALID_OBJECT_ID;
         if (m_int_ref1)

````

agrrr3 committed:
[focs] move some named_values to focs.py

Note from the error log (orthogonal to is this issue):
* XENOPHOBIC_MAX_JUMPS is set and only used as Integer - but for some reason it gets looked up as a real
** maybe there is generic lookup when printing the list of named values; maybe pass a flag that no specific type is expected which suppresses the info
* SetTopLevel warnings for the species could probably be removed by moving that definition into the central named_values

agrrr3 committed:
[focs] fix TopLevelContent overrides for weapons by calling NamedReal only once

* also pythonize reuse of pilot bonus

agrrr3 commented on parse/ValueRefPythonParser.cpp:
```diff
@@ -1299,6 +1347,8 @@ void RegisterGlobalsValueRefs(boost::python::dict& globals, const PythonParser&
     globals["JumpsBetween"] = insert_jumps_between_;
     globals["UserString"] = insert_user_string;
+    globals["NumPartClassesInShipDesign"] = boost::python::raw_function(insert_num_part_classes_in_ship_design_);
+    globals["PartOfClassInShipDesign"] = boost::python::raw_function(insert_part_of_class_in_ship_design_);
```
What is the difference between ValueRefPythonParser and ValueRefsPythonModuleParser?

agrrr3 committed:
Fix Incomplete Ship Design  PediaDetailPanel

In order for the current design to be queryable in a focs evaluation
it needs to be inserted into the context universe
so cost/turn evaluation has to happen below the call to
  universe.InsertShipDesignID

agrrr3 committed:
[pyi] Fix some _value_refs signatures

o01eg commented on universe/NamedValueRefManager.h:
```diff
@@ -231,6 +231,8 @@ class FO_COMMON_API NamedValueRefManager {
     FO_COMMON_API void SetNamedValueRefParse(Pending::Pending<NamedValueRefParseMap>&& future)
     { m_pending_named_value_refs_focs_txt = std::move(future); }
+    FO_COMMON_API void SetNamedValueRefPythonParse(Pending::Pending<NamedValueRefParseMap>&& future)
+    { m_pending_named_value_refs_focs_py = std::move(future); }
```
I would like to convert named values fast but I got my hands full.

agrrr3 commented on parse/ConditionPythonParser.cpp:
```diff
@@ -906,6 +935,7 @@ void RegisterGlobalsConditions(boost::python::dict& globals) {
     globals["OnPlanet"] = boost::python::raw_function(insert_on_planet_);
     globals["ResupplyableBy"] = boost::python::raw_function(insert_resupplyable_by_);
     globals["DesignHasPart"] = boost::python::raw_function(insert_design_has_part_);
+    globals["DesignHasPartClass"] = boost::python::raw_function(insert_design_has_part_class_);
```
`ConditionsPythonModuleParser.cpp` does not exist

agrrr3 committed:
[contents][focs][py][pyi] Policy DesignSimplicity based on count of part classes. Also *PartClass py support

* Policy DesignSimplicity gives a discount based on the complexity of the design.
* the more different ship part classes are, the lower the discount (with full cost with 5+ part classes)
* I.e. each part class adds one complexity, parts from colony parts add four complexity

Implementation:
* upkeep.py is not yet really usable from ship definitions, so upkeep.macros was used
* Add and use NumPartClassesInShipDesign vref for ship upkeep with PLC_DESIGN_SIMPLICITY.
* DESIGN_SIMPLICITY_SOURCE_COMPLEXITY_COUNT_VREF macro looks up vref of the same from named_values.focs.py
* the named vref uses NumPartClassesInShipDesign, for which only python parse is supported
* DESIGN_SIMPLICITY_COMPLEXITY_FACTOR_FOR_ARG1_VREF has a if-else cascade each rechecking the previous count and returning the discount
* stringtables
* AI (no changes necessary for basic support)

* Add python support for vref PartOfClassInShipDesign (in valueref module)
* Add python support for ShipPartClass enum
  Note that ship part class tokens have unique names like FuelClass to prevent name clash (e.g. with Fuel property)
* Add python support for condition DesignHasPartClass
* Add python signatures for NamedInteger,UsedInDesignID,NumPartClassesInShipDesign
* Streamline the python valueref module

What works:
* adjusted costs when single-click in ship designer UI
* adjusted costs in mouse over in build item picker
* adjusted costs in build queue
* adjusted costs when building items on turn
* adjusted costs when double-click/make a new ship design

Missing:
* performance testing

Notes
* both macros seem quite inefficient and have to be calculated for every ship design when opening the build queue. hoping caching/relying on invariants is good enough
* XXX maybe use a NamedValue for each design or store the complexity in the ship design itself in order to prevent recalculation
* part names are no objects; so this does not work (3 > Statistic UniqueCount value = LocalCandidate. candidate = CONDITION_RESULT_IS_OBJECTS
* ...cant discriminate part names ; there is no support for list type attributes

== AI ==
If it worked before, it should still work.

PolicyAI has a heuristic how good design simplicity is - that looks fine.

ShipDesigner should estimate ship design based on the real values with the applied policy,
so the simpler designs should be comparatively cheaper and the AI should use designs
based on efficiency afterwards.

The named values could be used in the AI, but are currently unused.

o01eg commented on parse/CMakeLists.txt:
```diff
@@ -95,6 +95,7 @@ if(BUILD_PARSERS)
             ${CMAKE_CURRENT_LIST_DIR}/EnumPythonParser.h
             ${CMAKE_CURRENT_LIST_DIR}/SourcePythonParser.h
             ${CMAKE_CURRENT_LIST_DIR}/ValueRefPythonParser.h
+            ${CMAKE_CURRENT_LIST_DIR}/ConditionsPythonModuleParser.h
```
.cpp file should be included too.

o01eg commented on parse/ConditionsPythonModuleParser.cpp:
```diff
@@ -7,13 +7,31 @@
 
+    template <typename T, typename... Args>
+    requires std::is_base_of_v<Condition::Condition, std::decay_t<T>>
+
```
Put them in a header?

agrrr3 replied:
i tried putting the template declarations in ConditionPythonParser.h and i could not make it work. I'll give up on it.

agrrr3 committed:
Include Condition.h before CommonParams.h gets included

agrrr3 commented:
will merge #5401 first and then rebase this on it

agrrr3 commented on parse/ValueRefsPythonModuleParser.cpp:
```diff
@@ -8,53 +8,53 @@
-            empire = pyobject_to_vref<int>(kw["empire"]);
+    value_ref_wrapper<int> insert_complex_i1_s1(std::string&& vname, const boost::python::tuple& args, const boost::python::dict& kw, const std::string&& keyint1, const std::string&& keystr1) {
+        std::unique_ptr<ValueRef::ValueRef<int>> int1;
+        if (kw.has_key(keyint1)) {
```
tracking this in #5409

agrrr3 commented on default/stringtables/en.txt:
```diff
@@ -12761,7 +12761,19 @@ PLC_DESIGN_SIMPLICITY
+Reduces production cost up to [[value PLC_DESIGN_SIMPLICITY_MAX_REDUCTION_PERCENT]]% for ships based on the ship design complexity. Not so much for outpost and colony ships.
+
+The final cost is
+• [[value PLC_DESIGN_SIMPLICITY_COMPLEXITY_FACTOR_AAA]] times the normal cost if there are no more than [[value PLC_DESIGN_SIMPLICITY_COMPLEXITY_MAX_COUNT_AAA]] different part classes.
```
Ok, I'll reword to match the content better
